Sources/AlgoliaSearch-Client/FilterBuilder/Filters/Filter.swift

```diff
@@ -0,0 +1,158 @@
+//
+
+import Foundation
+
+/// Abstract filter protocol
+public protocol Filter: Hashable {
+
+    /// Identifier of field affected by filter
+    var attribute: Attribute { get }
+    
+    /// A Boolean value indicating whether filter is inverted
+    var isInverted: Bool { get set }
+
+    /// String representation of filter excluding negation
+    var expression: String { get }
+    
+    /// Replaces isInverted property by a new value
+    /// parameter value: new value of isInverted
+    mutating func not(value: Bool)
+    
+    /// Returns string representation of filter
+    /// - parameter ignoringInversion: if set to true, ignores filter negation
+    func build(ignoringInversion: Bool) -> String
+    
+    /// Returns the same filter with attribute replaced by a provided one
+    /// - parameter attribute: attribute replacement
+    func replacingAttribute(by attribute: Attribute) -> Self
+}
+
+private class _AnyFilterBase: Filter {
+    
+    var attribute: Attribute {
+        fatalError("Must override")
+    }
+    
+    var isInverted: Bool {
+        get { fatalError("Must override") }
+        
+        set { fatalError("Must override") }
+    }
+    
+    var expression: String {
+        fatalError("Must override")
+    }
+    
+    var hashValue: Int {
+        fatalError("Must override")
+    }
+    
+    init() {
+        guard type(of: self) != _AnyFilterBase.self else {
+            fatalError("_AnyFilterBase instances can not be created; create a subclass instance instead")
+        }
+    }
+    
+    func replacingAttribute(by attribute: Attribute) -> Self {
+        fatalError("Must override")
+    }
+    
+    static func == (lhs: _AnyFilterBase, rhs: _AnyFilterBase) -> Bool {
+        fatalError("Must override")
+    }
+    
+}
+
+private final class _AnyFilterBox<Concrete: Filter>: _AnyFilterBase {
+
+    var concrete: Concrete
+    
+    init(_ concrete: Concrete) {
+        self.concrete = concrete
+    }
+    
+    override var attribute: Attribute {
+        return concrete.attribute
+    }
+    
+    override var isInverted: Bool {
+        get { return concrete.isInverted }
+        set { concrete.isInverted = newValue }
+    }
+    
+    override var expression: String {
+        return concrete.expression
+    }
+    
+    override var hashValue: Int {
+        return concrete.hashValue
+    }
+    
+    override func replacingAttribute(by attribute: Attribute) -> _AnyFilterBox {
+        return _AnyFilterBox(concrete.replacingAttribute(by: attribute))
+    }
+    
+    static func == (lhs: _AnyFilterBox, rhs: _AnyFilterBox) -> Bool {
+        return lhs.concrete == rhs.concrete
+    }
+    
+}
+
+final class AnyFilter: Filter {
+    
+    private let box: _AnyFilterBase
+    
+    init<Concrete: Filter>(_ concrete: Concrete) {
+        box = _AnyFilterBox(concrete)
+    }
+    
+    var attribute: Attribute {
+        return box.attribute
+    }
+    
+    var isInverted: Bool {
```

Comment:
I would love to call this `isNegated`. It is more accurate, as we either add a `NOT` keyword before the filter, or we don't.

Reply:
Done in https://github.com/algolia/algoliasearch-client-swift/pull/597/commits/0be30dd5e0a282449ccc49e88a83bf6101377ade